File: Stack-Queue-LinkedList/876. Middle of the Linked List/Solution.java
/**
 Given a non-empty, singly linked list with head node head, return a middle node of linked list.
 If there are two middle nodes, return the second middle node.
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        
        ListNode slow = head;
        ListNode fast = head;
        
        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        return slow;
    }
}

/**
 If there are two middle nodes, return the 1st middle node.
 */

class Solution {
  public ListNode middleNode(ListNode head) {
      
      ListNode slow = head;
      ListNode fast = head;
      
      while(fast.next != null && fast.next.next != null) {
          slow = slow.next;
          fast = fast.next.next;
      }
      
      return slow;
  }
}

File: Stack-Queue-LinkedList/876. Middle of the Linked List/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# return second node if there are two mid (even length)
class Solution(object):
    def middleNode(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """

        if not head or not head.next:
            return head

        slow = node
        fast = node
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

# return 1st node if there are two mid (even length)
class Solution(object):
    def middleNode(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """

        if not head or not head.next:
            return head

        slow = node
        fast = node
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        return slow